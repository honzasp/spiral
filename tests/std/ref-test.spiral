(program
  (import std)
  (import std.test)
  (import std.array)

  (test "ref?" (lambda (t)
    (assert t (ref? (ref-new 99)))
    (assert-not t (ref? (sym-new "a sym")))
    (assert-not t (ref? "not a ref"))))

  (test "ref-get" (lambda (t)
    (let ((r0 (ref-new 1000))
          (r1 (ref-new "baz")))
      (assert-eqv t (ref-get r0) 1000)
      (assert-eqv t (ref-get r1) "baz"))))

  (test "ref-set!" (lambda (t)
    (let ((r0 (ref-new "baz"))
          (ary0 (array-new))
          (ary1 (array-new))
          (r1 (ref-new ary0))
          (r2 (ref-new ary0)))
      (assert-eqv t (ref-get r0) "baz")
      (assert-eqv t (ref-get r1) (ref-get r2))
      (ref-set! r0 "foo")
      (assert-eqv t (ref-get r0) "foo")
      (ref-set! r0 1234)
      (assert-eqv t (ref-get r0) 1234)
      (ref-set! r2 ary1)
      (assert-eqv t (ref-get r1) ary0)
      (assert-eqv t (ref-get r2) ary1)))))
