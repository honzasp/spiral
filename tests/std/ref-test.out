test ref-get ok
test ref-set! ok
